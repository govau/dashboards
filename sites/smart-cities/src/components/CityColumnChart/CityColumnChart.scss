@import '~style/main';

.wrapper {
  padding-top: 16px;
  margin-bottom: 48px;
}

.titleWrapper {
  display: flex;
  align-items: center;
  line-height: 1.2;
  margin-bottom: 3px;
}

.title {
  font-size: 22px;
  font-weight: 300;
  margin-right: 8px;

  @include for-tablet-landscape-up {
    font-size: 24px;
  }
}

.metaWrapper {
  font-size: 12px;
  margin-bottom: 4px;
}

.indicatorTypeMark {
  margin: -3px 16px 0 0;
}

.legendWrapper {
  display: flex;
  flex-flow: wrap;
  justify-content: flex-end;
  margin-bottom: 8px;
}

.axisTitle {
  height: 20px; // we want it to take up space even with no text
  margin-top: 16px;
  font-size: 12px;
}

.chartWrapper {
  min-height: 400px; // so when the charts are loading it doesn't collapse

  @include for-tablet-landscape-up {
    min-height: 500px;
  }
}

.aboutChartIcon {
  margin-top: 3px;
}

.table {
  @include hidden;
}

:global(.cityColumnChartTooltipHeader) {
  margin-left: 8px;
  margin-right: 8px;
  margin-bottom: 6px;
  font-size: 12px;
  font-weight: 500;
}

:global(.cityColumnChartTooltipRow) {
  margin-left: 8px;
  margin-right: 8px;
  font-size: 12px;

  &:not(:last-of-type) {
    margin-bottom: 4px;
  }
}

:global(.cityColumnChartTooltipDot) {
  position: relative;
  top: 2px;
  margin-right: 4px;
}

:global(.cityColumnChartTooltipValue) {
  font-weight: 700;
}
