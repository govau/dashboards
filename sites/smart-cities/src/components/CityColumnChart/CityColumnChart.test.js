import React from 'react';
import { shallow } from 'enzyme';
import CityColumnChart from './CityColumnChart';
import Highcharts from 'highcharts';

const chartMock = {
  destroy: jest.fn(),
};

Highcharts.chart = jest.fn(() => chartMock);

console.warn = jest.fn();

jest.mock('../Tooltip/Tooltip', () => 'Tooltip');
jest.mock('../Legend/Legend', () => 'Legend');
jest.mock('./ChartTable/ChartTable', () => 'ChartTable');
jest.mock('../Icon/Icon', () => 'Icon');
jest.mock('../IndicatorIcon/IndicatorIcon', () => 'IndicatorIcon');
jest.mock('../../constants', () => ({
  STRINGS: {
    CONTEXTUAL_DEFINITION: 'CONTEXTUAL_DEFINITION',
    PERFORMANCE_DEFINITION: 'PERFORMANCE_DEFINITION',
  },
  COLOR_NAMES: {
    JOBS: 'jobs',
    PLANNING: 'planning',
  },
  INDICATORS: {
    population: {
      name: 'The population indicator name',
      format: '0[.]0a',
      contextual: true,
    },
    growthRate: {
      name: 'The growthRate indicator name',
      format: '0[.]0a',
      contextual: false,
    },
  }
}));

jest.mock('../../helpers/getColorRange', () => () => [
  'red',
  'yellow',
  'blue',
]);

const defaultProps = {
  chart: {
    indicatorIds: [
      'population'
    ],
    description: 'the description of the task',
  },
  colorBase: 'jobs',
  highlightColorLight: 'JOBS_040',
  highlightColorDark: 'JOBS_600',
  cities: [
    {
      name: 'Perth',
      indicators: {
        population: 12,
        growthRate: 0.99,
      }
    },
    {
      name: 'Sydney',
      indicators: {
        population: 12,
        growthRate: 0.19,
      }
    },
  ],
  showChartPatterns: false,
  svgPatternBase: 'svg-pattern-base',
};

it('should match Snapshot', () => {
  const component = shallow(
    <CityColumnChart {...defaultProps} />
  );

  expect(component.debug()).toMatchSnapshot();
});

it('should not render the chart immediately', () => {
  jest.useFakeTimers();

  shallow(<CityColumnChart {...defaultProps} />);

  // not rendered yet
  expect(Highcharts.chart).toHaveBeenCalledTimes(0);

  // fast forward until all timers are fired
  jest.runAllTimers();

  // now it's rendered
  expect(Highcharts.chart).toHaveBeenCalledTimes(1);
});

it('should call highcharts with the right div ID', () => {
  jest.useFakeTimers();

  shallow(<CityColumnChart {...defaultProps} />);

  jest.runAllTimers();

  expect(Highcharts.chart).toHaveBeenCalledWith(
    'population-chart',
    expect.any(Object),
  );
});

it('should call highcharts with the right config', () => {
  jest.useFakeTimers();

  shallow(<CityColumnChart {...defaultProps} />);

  jest.runAllTimers();

  const config = Highcharts.chart.mock.calls[0][1];

  expect(config.series.length).toBe(1);
  expect(config.xAxis.categories.length).toBe(2); // the two mock cites
  expect(config.tooltip.borderColor).toBe('JOBS_900');
});

it('should call highcharts with the right config for multiple indicators', () => {
  jest.useFakeTimers();

  shallow(
    <CityColumnChart
      {...defaultProps}
      chart={{
        description: 'the description',
        indicatorIds: [
          'population',
          'growthRate',
        ]
      }}
    />);

  jest.runAllTimers();

  const config = Highcharts.chart.mock.calls[0][1];

  expect(config.series.length).toBe(2); // the two indicators
});

it('should call highcharts with the right config for showing patterns', () => {
  jest.useFakeTimers();

  // only show patterns when there's multiple indicators and showChartPatterns is true
  shallow(
    <CityColumnChart
      {...defaultProps}
      chart={{
        description: 'the description',
        indicatorIds: [
          'population',
          'growthRate',
        ]
      }}
      showChartPatterns={true}
    />);

  jest.runAllTimers();

  const config = Highcharts.chart.mock.calls[0][1];

  expect(config.series[0].color).toBe('url(#svg-pattern-base-0)');
  expect(config.series[1].color).toBe('url(#svg-pattern-base-1)');
});

it('should call highcharts with the right config for not showing patterns', () => {
  jest.useFakeTimers();

  // only show patterns when there's multiple indicators and showChartPatterns is true
  shallow(
    <CityColumnChart
      {...defaultProps}
      chart={{
        description: 'the description',
        indicatorIds: [
          'population',
          'growthRate',
        ]
      }}
      showChartPatterns={false}
    />);

  jest.runAllTimers();

  const config = Highcharts.chart.mock.calls[0][1];

  expect(config.series[0].color).toBe('red');
  expect(config.series[1].color).toBe('yellow');
});

it('should update the chart when new props arrive', () => {
  jest.useFakeTimers();

  const component = shallow(<CityColumnChart {...defaultProps} />);

  jest.runAllTimers();

  expect(Highcharts.chart).toHaveBeenCalledTimes(1);

  component.setProps({
    colorBase: 'planning',
  });

  // props changed, and a chart update is queued, but not run yet
  expect(Highcharts.chart).toHaveBeenCalledTimes(1);

  // now run the timer
  jest.runAllTimers();

  // now it's been updated (rendered again)
  expect(Highcharts.chart).toHaveBeenCalledTimes(2);

  const config = Highcharts.chart.mock.calls[1][1];

  expect(config.tooltip.borderColor).toBe('PLANNING_900');
});

it('should destroy the chart when dismounting', () => {
  jest.useFakeTimers();

  const component = shallow(<CityColumnChart {...defaultProps} />);

  jest.runAllTimers();

  expect(chartMock.destroy).toHaveBeenCalledTimes(0);

  component.unmount();

  expect(chartMock.destroy).toHaveBeenCalledTimes(1);
});

it('should render a legend when there is more than one indicator', () => {
  const component = shallow(
    <CityColumnChart
      {...defaultProps}
      chart={{
        description: 'the description',
        indicatorIds: [
          'population',
          'growthRate',
        ]
      }}
    />
  );

  expect(component.find('Legend').length).toBe(1);
});

it('should not render a legend when there is only one indicator', () => {
  const component = shallow(
    <CityColumnChart
      {...defaultProps}
    />
  );

  expect(component.find('Legend').length).toBe(0);
});
