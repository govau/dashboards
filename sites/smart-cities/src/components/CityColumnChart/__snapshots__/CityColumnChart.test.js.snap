// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match Snapshot 1`] = `
"<article className=\\"wrapper\\">
  <ChartTable className=\\"table\\" chart={{...}} cities={{...}} />
  <div aria-hidden=\\"true\\">
    <header className=\\"titleWrapper\\">
      <h1 className=\\"title\\" />
      <Tooltip text=\\"the description of the task\\" borderColor=\\"JOBS_600\\">
        <Icon icon=\\"questionMark\\" className=\\"aboutChartIcon\\" size={22} color=\\"JOBS_600\\" title=\\"\\" />
      </Tooltip>
    </header>
    <div className=\\"metaWrapper\\">
      <IndicatorIcon className=\\"indicatorTypeMark\\" color=\\"JOBS_500\\" contextual={true} />
    </div>
    <div className=\\"axisTitle\\" />
    <div className=\\"chartWrapper\\">
      <div id=\\"population-chart\\" />
    </div>
  </div>
</article>"
`;
